### PR TITLE
add middleware tests and change middleware to filter up to down

### PR DESCRIPTION
**Pull request by chenshoo**

### CODE REVIEW

File: src/middlewares/filter-resolver.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { isRepositoryEntity } from '../dal/entities/repository-entity';
+import { MiddlewaresConfiguration, ResponseMiddlewareParams } from './middleware';
+import { DataVersionFilter } from './middleware-activation-condition/filter-data-version';
+import { RealityIdFilter } from './middleware-activation-condition/filter-realities';
+import { SoftDeleteFilter } from './middleware-activation-condition/filter-soft-delete';
+
+export class FilterResolver {
```
Comment (osher-sade) [inline src/middlewares/filter-resolver.ts]: Rename to FilterExecutor

File: src/middlewares/filter-resolver.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    middlewaresConfig: MiddlewaresConfiguration;
+
+    constructor(middlewaresConfig?: MiddlewaresConfiguration) {
```
Comment (osher-sade) [inline src/middlewares/filter-resolver.ts]: Make middleware config and all of it's properties to be mandatory

File: src/middlewares/filter-resolver.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        this.middlewaresConfig = middlewaresConfig
+            ? middlewaresConfig
+            : { allowDataVersionMiddleware: true, allowRealityMiddleware: true };
+        if (this.middlewaresConfig.allowRealityMiddleware !== false) {
+            this.middlewaresConfig.allowRealityMiddleware = true;
+        }
+    }
+
+    filterResolveResult(params: ResponseMiddlewareParams): any {
```
Comment (osher-sade) [inline src/middlewares/filter-resolver.ts]: rename to executeFilters

File: src/middlewares/filter-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    arrangeDataAndFilter(params: ResponseMiddlewareParams): any {
+        const results: any = [];
+        for (const entity of params.result as any) {
+            results.push(entity._doc);
+        }
+        params.result = results;
+        return this.filterEntities(params);
+    }
+
+    filterEntities(params: ResponseMiddlewareParams): any {
```
Comment (osher-sade) [inline src/middlewares/filter-resolver.ts]: rename to filterRootEntities

File: src/middlewares/filter-resolver.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        const res: any = [];
+        for (const entity of params.result as any) {
+            params.result = entity;
+            if (
+                isRepositoryEntity(entity) &&
```
Comment (osher-sade) [inline src/middlewares/filter-resolver.ts]: extract to boolean function

File: src/middlewares/polaris-middleware.ts [hunk 6]
```diff
@@ -1,23 +1,19 @@
...
 import { Middleware, RequestMiddlewareParams, ResponseMiddlewareParams } from './middleware';
-import { DataVersionFilter } from './middleware-activation-condition/filter-data-version';
-import { RealityIdFilter } from './middleware-activation-condition/filter-realities';
-import { SoftDeleteFilter } from './middleware-activation-condition/filter-soft-delete';
 
 @injectable()
 export class PolarisMiddleware implements Middleware {
     @inject(POLARIS_TYPES.GraphqlLogger) polarisLogger!: GraphqlLogger<PolarisContext>;
-    middlewaresConfig: MiddlewaresConfiguration;
+    filterResolver: FilterExecutor;
```
Comment (osher-sade) [inline src/middlewares/polaris-middleware.ts]: Change variable name

File: src/middlewares/polaris-middleware.ts [hunk 7]
```diff
@@ -1,23 +1,19 @@
...
 
     constructor(@inject(POLARIS_TYPES.MiddlewaresConfig) middlewaresConfig: MiddlewaresConfig) {
-        this.middlewaresConfig = middlewaresConfig.middlewaresConfiguration;
+        this.filterResolver = new FilterExecutor(middlewaresConfig.middlewaresConfiguration);
     }
 
     preResolve({ root, info, context, args }: RequestMiddlewareParams): void {
```
Comment (osher-sade) [inline src/middlewares/polaris-middleware.ts]: Create a method for logStartOfResolve as well (like you did with logEndOfResolve)

File: src/middlewares/filter-executor.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+export class FilterExecutor {
+    middlewaresConfig: MiddlewaresConfiguration;
+
+    constructor(middlewaresConfig: MiddlewaresConfiguration) {
+        this.middlewaresConfig = middlewaresConfig;
+    }
+
+    executeFilters(params: ResponseMiddlewareParams): any {
+        return params.root ? this.filterSubEntity(params) : this.filterRootEntities(params);
```
Comment (yarinvak) [inline src/middlewares/filter-executor.ts]: if root filter sub entity else filter root?

File: src/middlewares/filter-executor.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        const notRepositoryEntities: any[] = [];
+        for (const entity of params.result as any) {
+            if (isRepositoryEntity(entity._doc)) {
+                repositoryEntities.push(entity._doc);
+            } else {
+                notRepositoryEntities.push(entity);
+            }
+        }
+        params.result = repositoryEntities;
+        return notRepositoryEntities.concat(this.filterRootRepositoryEntities(params));
```
Comment (yarinvak) [inline src/middlewares/filter-executor.ts]: why sending the params again if you already hold the repository entities list?

File: src/middlewares/filter-executor.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
+import { isRepositoryEntity } from '../dal/entities/repository-entity';
+import { MiddlewaresConfiguration, ResponseMiddlewareParams } from './middleware';
+import { DataVersionFilter } from './middleware-activation-condition/filter-data-version';
+import { RealityIdFilter } from './middleware-activation-condition/filter-realities';
+import { SoftDeleteFilter } from './middleware-activation-condition/filter-soft-delete';
+
+export class FilterExecutor {
```
Comment (yarinvak) [inline src/middlewares/filter-executor.ts]: why so many methods?

File: src/middlewares/middleware-activation-condition/filter-data-version.ts [hunk 11]
```diff
@@ -1,10 +1,13 @@
 import { ResponseMiddlewareParams } from '../middleware';
-import { isSubEntity, MiddlewareCondition } from './filter-condition';
+import { MiddlewareCondition } from './filter-condition';
 
 class FilterDataVersion implements MiddlewareCondition {
-    shouldBeReturned({ root, context, info }: ResponseMiddlewareParams): boolean {
+    shouldBeReturned(
+        { root, context, result }: ResponseMiddlewareParams,
+        subEntity: boolean,
```
Comment (yarinvak) [inline src/middlewares/middleware-activation-condition/filter-data-version.ts]: isSubEntity

File: src/middlewares/middleware-activation-condition/filter-realities.ts [hunk 12]
```diff
@@ -1,19 +1,26 @@
...
 import { ResponseMiddlewareParams } from '../middleware';
-import { isSubEntity, MiddlewareCondition } from './filter-condition';
+import { MiddlewareCondition } from './filter-condition';
 
 class FilterRealities implements MiddlewareCondition {
-    shouldBeReturned({ root, context, info }: ResponseMiddlewareParams): boolean {
-        const operationalRealityId: string = '0';
+    shouldBeReturned(
+        { root, context, info, result }: ResponseMiddlewareParams,
+        subEntity: boolean,
```
Comment (yarinvak) [inline src/middlewares/middleware-activation-condition/filter-realities.ts]: isSubEntity

File: src/middlewares/middleware-activation-condition/filter-realities.ts [hunk 13]
```diff
@@ -1,19 +1,26 @@
...
+    ): boolean {
+        const operationalRealityId: number = 0;
         const headers: PolarisRequestHeaders = context.headers;
-        const isMatchingRealities: boolean = root.realityId === headers.realityId;
+        const realityIdHeaderExists: any = headers.realityId === 0 ? true : headers.realityId;
+        const isMatchingRealities: boolean = subEntity
```
Comment (yarinvak) [inline src/middlewares/middleware-activation-condition/filter-realities.ts]: isSubEntity

File: src/middlewares/middleware-activation-condition/filter-realities.ts [hunk 14]
```diff
@@ -1,19 +1,26 @@
...
+            ? root.realityId === headers.realityId
+            : result.realityId === headers.realityId;
         return (
-            !headers.realityId ||
+            !realityIdHeaderExists ||
             isMatchingRealities ||
-            (isSubEntity(info) &&
+            (subEntity &&
```
Comment (yarinvak) [inline src/middlewares/middleware-activation-condition/filter-realities.ts]: isSubEntity

**Pull request merged**